### PR TITLE
Encapsulate Project

### PR DESCRIPTION
**Pull request by rodiguif**
This PR Creates a simulation of forecast chart and encapsulates the project with a return from anvil view to use as a library.

forecastchart

### CODE REVIEW

File: app/src/main/java/com/oxeanbits/forecastchart/model/LineDataCreator.kt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package com.oxeanbits.forecastchart.model
+
+import com.github.mikephil.charting.data.Entry
+
+data class LineDataCreator (
```
Comment (tiagogoncalves) [inline app/src/main/java/com/oxeanbits/forecastchart/model/LineDataCreator.kt]:
```suggestion
data class LineData (
```

**or** 

```suggestion
data class Line (
```

File: app/src/main/java/com/oxeanbits/forecastchart/model/LineDataCreator.kt [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+    val values: ArrayList<Entry>,
```
Comment (tiagogoncalves) [inline app/src/main/java/com/oxeanbits/forecastchart/model/LineDataCreator.kt]: `Entry` is a pair, right? What type we can pass? String, Decimal, Integer?

File: app/src/main/java/com/oxeanbits/forecastchart/util/SetupChart.kt [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+import com.github.mikephil.charting.data.LineDataSet
+import com.github.mikephil.charting.utils.ColorTemplate
+import java.text.SimpleDateFormat
+import com.oxeanbits.forecastchart.R
+import com.oxeanbits.forecastchart.ui.marker.ForecastMarkerView
+import com.oxeanbits.forecastchart.model.LineDataCreator
+
+object SetupChart{
+
+    fun setupForecastChart(context: Context, combinedChart: CombinedChart, l1: LineDataCreator, l2: LineDataCreator) {
```
Comment (tiagogoncalves) [inline app/src/main/java/com/oxeanbits/forecastchart/util/SetupChart.kt]: We have the practice to have at maximum 100 characters per line.

File: app/src/main/java/com/oxeanbits/forecastchart/util/SetupChart.kt [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        val actual = setupLineDataSet(l1.values, l1.label, l1.color)
+        val production = setupLineDataSet(l2.values, l2.label, l2.color)
+        val endDateArray = arrayListOf(BarEntry(l2.values.get(l2.values.size-1).x, l2.values.get(l2.values.size-1).y))
```
Comment (tiagogoncalves) [inline app/src/main/java/com/oxeanbits/forecastchart/util/SetupChart.kt]: Same here.

File: app/src/main/java/com/oxeanbits/forecastchart/util/SetupChart.kt [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        lineDataSet.color = color
+        lineDataSet.lineWidth = 2F
+        lineDataSet.setDrawCircles(false)
+        lineDataSet.setDrawValues(false)
+        lineDataSet.setDrawFilled(true)
+        lineDataSet.fillColor = color
+        lineDataSet.fillAlpha = 45
+
+        return lineDataSet
+    }
```
Comment (tiagogoncalves) [inline app/src/main/java/com/oxeanbits/forecastchart/util/SetupChart.kt]: Let's have this in the test module

**Review by tiagogoncalves — CHANGES_REQUESTED** [review]
Really nice, man. I did not have time to run this, but for now I have some questions.

**Comment by tiagogoncalves** [comment]
I thing that I saw in your PR is you are creating the project and doing the things here. In the future let's create more defined PRs, ok?

File: app/src/main/java/com/oxeanbits/forecastchart/model/LineDataCreator.kt (hunk 2 above)
Comment (rodiguif) [reply app/src/main/java/com/oxeanbits/forecastchart/model/LineDataCreator.kt]: `Entry` is a pair of Floats (X and Y axis).

File: app/src/main/java/com/oxeanbits/forecastchart/model/LineDataCreator.kt (hunk 1 above)
Comment (rodiguif) [reply app/src/main/java/com/oxeanbits/forecastchart/model/LineDataCreator.kt]: `LineData` is already a chart library model. I think that can be `Line`.

File: .idea/runConfigurations.xml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (niltonvasques) [inline .idea/runConfigurations.xml]: These `.idea` files are personal and they don't need to be tracked by git.

File: app/.gitignore [hunk 7]
```diff
@@ -0,0 +1 @@
+/build
```
Comment (niltonvasques) [inline app/.gitignore:1]:
```suggestion
/build
.idea
```

File: app/build.gradle [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    implementation 'androidx.appcompat:appcompat:1.1.0'
+    implementation 'androidx.core:core-ktx:1.1.0'
+
+    implementation 'co.trikita:anvil-sdk15:0.5.2'
+    implementation 'co.trikita:anvil-design:0.5.2'
+
+    implementation 'com.github.PhilJay:MPAndroidChart:v3.1.0'
+
+    testImplementation "org.mockito:mockito-core:2.23.0"
+    testImplementation 'com.squareup.okhttp3:mockwebserver:3.6.0'
```
Comment (niltonvasques) [inline app/build.gradle]: I believe that this dependency is not necessary here

File: app/src/main/java/com/oxeanbits/forecastchart/util/ForecastChart.kt [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+object ForecastChart{
+
+    fun createForecastChart(context: Context, combinedChart: CombinedChart, l1: Line, l2: Line) {
+        val actual = setupLineDataSet(l1.values, l1.label, l1.color)
+        val production = setupLineDataSet(l2.values, l2.label, l2.color)
+
+        val endX = l2.values.get(l2.values.size-1).x
+        val endY = l2.values.get(l2.values.size-1).y
+        val endDateArray = arrayListOf(BarEntry(endX, endY))
+        val endDate = setupEndBarDataSet(endDateArray, "End Date", Color.RED)
```
Comment (niltonvasques) [inline app/src/main/java/com/oxeanbits/forecastchart/util/ForecastChart.kt]: Would be possible to extract this `End Date` locale to a resource?

File: sync/src/main/java/com/oxeanbits/sync/ui/ForecastMarkerView.kt [hunk 10]
```diff
@@ -0,0 +1,39 @@
+package com.oxeanbits.sync.ui
```
Comment (niltonvasques) [inline sync/src/main/java/com/oxeanbits/sync/ui/ForecastMarkerView.kt]: Is this module used?

**Review by niltonvasques — CHANGES_REQUESTED** [review]
What about to make this component more easy to use. 

The ideal interface would looks like that:

```kotlin
linearLayout {
   width(MATCH_PARENT)
   height(MATCH_PARENT)

  forecastChart {
     width(MATCH_PARENT)
     height(MATCH_PARENT)
     expectedData(expected)
     actualData(actual)
  }
}
```

File: sync/src/main/java/com/oxeanbits/sync/ui/ForecastMarkerView.kt (hunk 10 above)
Comment (rodiguif) [reply sync/src/main/java/com/oxeanbits/sync/ui/ForecastMarkerView.kt]: Yes, the `sync` module is used by the `app` and `example` module. And its `ui` is the point detail view when the chart is pressed.

Comment (niltonvasques) [reply sync/src/main/java/com/oxeanbits/sync/ui/ForecastMarkerView.kt]: In this case we need to rename its package to:

```suggestion
package com.oxeanbits.forecastchart.sync.ui
```

File: core/src/main/java/com/oxeanbits/forecastchart/core/ui/component/CombinedChartComponent.kt [hunk 11]
```diff
@@ -31,17 +41,24 @@ class CombinedChartComponent(context: Context) : LinearLayout(context), Anvil.Re
             v(CombinedChart::class.java) {
                 size(MATCH, 500)
                 margin(10)
-                init{
-                    combinedChart = Anvil.currentView()
+                init {
+                    this.combinedChart = Anvil.currentView()
+                    this.combinedChart?.let {
```
Comment (niltonvasques) [inline core/src/main/java/com/oxeanbits/forecastchart/core/ui/component/CombinedChartComponent.kt:46]: An alternative version, solving the anti arrow pattern:

```suggestion
val combinedChart =  this.combinedChart ?: return
val expectedData = this.expectedData ?: return
val actualData = this.actualData ?: return
ForecastChart.createForecastChart(context, combinedChart, expectedData, actualData)
```

**Review by niltonvasques — APPROVED** [review]
Lets open another PR with the refactors @rodiguif

**Pull request merged**